src/main/java/spoon/reflect/visitor/PrinterTokenWriter.java

```diff
@@ -0,0 +1,46 @@
+/**
+ * Copyright (C) 2006-2017 INRIA and contributors
+ * Spoon - http://spoon.gforge.inria.fr/
+ *
+ * This software is governed by the CeCILL-C License under French law and
+ * abiding by the rules of distribution of free software. You can use, modify
+ * and/or redistribute the software under the terms of the CeCILL-C license as
+ * circulated by CEA, CNRS and INRIA at http://www.cecill.info.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE. See the CeCILL-C License for more details.
+ *
+ * The fact that you are presently reading this means that you have had
+ * knowledge of the CeCILL-C license and that you accept its terms.
+ */
+package spoon.reflect.visitor;
+
+import spoon.reflect.visitor.printer.TokenType;
+
+/**
+ * The token based and context aware printer.
+ */
+public interface PrinterTokenWriter {
+	/**
+	 * Writes one token of printed java source code to the registered {@link PrinterHelper}
+	 * @param type The type of the token
+	 * @param token the value of the token. The value type depends on TokenType. See {@link TokenType#getTokenClass()}
+	 */
+	void writeToken(TokenType type, Object token);
```

Comment:
Typing by Object is rarely good.  Any way to have a more specific type?

Reply:
token value has actually these types:
* String
* CtComment
* CtJavaDocTag

I do not know yet how to avoid Object. May be later ... Do you have any idea?

Reply:
one alternative would be that instead of having the additional enum, to have a visitor like design with three methods
writeString(String), writeComment(CtComment), visitJavaDocTag(CtJavaDocTag)